File: frontend/src/types/operational.ts
export type DataQuality = "VALID" | "WARNING" | "DIVERGENT" | "INVALID";

export interface PageMeta {
  page: number;
  page_size: number;
  total: number;
  pages: number;
}

export interface QualityMessage {
  type: string;
  severity: "WARNING" | "DIVERGENT";
  message: string;
}

export interface SaleItem {
  id: string;
  contract_code: string | null;
  client_name: string | null;
  client_identity_id?: number | null;
  client_name_source?: "CLIENT_CANONICAL" | "ECON_EMPRESTIMOS" | null;
  client_name_divergent?: boolean;
  source_client_code: string | null;
  operation_date: string | null;
  release_date: string | null;
  first_due_date: string | null;
  term: number | null;
  principal: string | null;
  iof: string | null;
  financed_amount: string | null;
  installment_amount: string | null;
  released_amount: string | null;
  interest_rate: string | null;
  irr_rate: string | null;
  cet_monthly_rate: string | null;
  status: string | null;
  data_quality_status: DataQuality;
  warning_count: number;
  divergence_count: number;
  funding_status: "NOT_INFORMED" | "INCOMPLETE" | "COMPLETE" | "OVERFUNDED" | "BASE_AMOUNT_UNAVAILABLE";
  funding_identified_amount: string;
  funding_difference: string | null;
  funding_source_count: number;
  bank_validation_status: "NOT_RECORDED" | "VALIDATED" | "DIVERGENT";
  continuity_id: string | null;
  continuity_type: "REFINANCING" | "RENEGOTIATION" | "ROLLOVER" | null;
  continuity_role: "PREDECESSOR" | "SUCCESSOR" | null;
  predecessor_sale_id: string | null;
  predecessor_contract_code: string | null;
  predecessor_sale_ids?: string[];
  predecessor_contract_codes?: string[];
  successor_sale_id: string | null;
  successor_contract_code: string | null;
  continuity_effective_date: string | null;
  continuity_notes: string | null;
}

export interface SaleDetail extends SaleItem {
  warnings: QualityMessage[];
  divergences: QualityMessage[];
}

export interface SalesResponse {
  items: SaleItem[];
  pagination: PageMeta;
  summary: {
    total_contracts: number;
    principal: string;
    released_amount: string;
    financed_amount: string;
    warning_contracts: number;
    divergent_contracts: number;
  };
}

export interface RevenueItem {
  id: number;
  revenue_identity_id?: string | null;
  contract_code: string | null;
  client_name: string | null;
  client_name_source?: "CLIENT_CANONICAL" | "ECON_EMPRESTIMOS" | null;
  client_name_divergent?: boolean;
  installment_code: string | null;
  due_date: string | null;
  payment_date: string | null;
  expected_amount: string | null;
  paid_amount: string | null;
  principal_component: string | null;
  interest_component: string | null;
  discount_amount: string | null;
  installment_status: string | null;
  situation: string | null;
  anticipation_marker: string | null;
  data_quality_status: DataQuality;
  warning_count: number;
  divergence_count: number;
  sale_id: string | null;
  funding_status: SaleItem["funding_status"] | null;
  distribution_status: "PENDING_FUNDING" | "READY" | "DISTRIBUTED" | "DIVERGENT" | "REVERSED";
  primary_source_name: string | null;
  bank_validation_status: "NOT_RECORDED" | "VALIDATED" | "DIVERGENT";
  refinanced_to_contract_code: string | null;
}

export interface RevenueDetail extends RevenueItem {
  payment_marker: string | null;
  source_reference: string | null;
  warnings: QualityMessage[];
  divergences: QualityMessage[];
  bank_validation_status: "NOT_RECORDED" | "VALIDATED" | "DIVERGENT";
}

export interface RevenueResponse {
  items: RevenueItem[];
  pagination: PageMeta;
  summary: {
    total_records: number;
    expected_amount: string;
    paid_amount: string;
    principal_received: string;
    interest_amount: string;
    discount_amount: string;
    pending_records: number;
    warning_records: number;
    divergent_records: number;
    principal_total: string;
    principal_open: string;
    average_pmt: string;
    overdue_principal: string;
    overdue_pmt: string;
    delinquency_percentage: string;
  };
}

export interface SalesFilters {
  page?: number;
  page_size?: number;
  search?: string;
  contract?: string;
  client?: string;
  status?: string;
  period_from?: string;
  period_to?: string;
  quality?: DataQuality | "";
  sort_by?: string;
  sort_order?: "asc" | "desc";
}

export interface RevenueFilters {
  page?: number;
  page_size?: number;
  search?: string;
  contract?: string;
  client?: string;
  status?: string;
  due_from?: string;
  due_to?: string;
  payment_from?: string;
  payment_to?: string;
  quality?: DataQuality | "";
  view?: "all" | "received" | "open" | "overdue" | "future";
  sort_by?: string;
  sort_order?: "asc" | "desc";
}

File: backend/app/schemas/operational.py
from __future__ import annotations

from datetime import date
from decimal import Decimal
from typing import Annotated, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, PlainSerializer

Money = Annotated[Decimal, PlainSerializer(lambda value: format(value, "f"), return_type=str)]
Quality = Literal["VALID", "WARNING", "DIVERGENT", "INVALID"]


class OperationalApiModel(BaseModel):
    model_config = ConfigDict(from_attributes=True)


class PageMeta(OperationalApiModel):
    page: int
    page_size: int
    total: int
    pages: int


class QualityMessage(OperationalApiModel):
    type: str
    severity: str
    message: str


class SaleSummary(OperationalApiModel):
    total_contracts: int
    principal: Money
    released_amount: Money
    financed_amount: Money
    warning_contracts: int
    divergent_contracts: int


class SaleItem(OperationalApiModel):
    id: str
    contract_code: str | None
    client_name: str | None
    client_identity_id: int | None = None
    client_name_source: Literal["CLIENT_CANONICAL", "ECON_EMPRESTIMOS"] | None = None
    client_name_divergent: bool = False
    source_client_code: str | None
    operation_date: date | None
    release_date: date | None
    first_due_date: date | None
    term: int | None
    principal: Money | None
    iof: Money | None
    financed_amount: Money | None
    installment_amount: Money | None
    released_amount: Money | None
    interest_rate: Money | None
    irr_rate: Money | None
    cet_monthly_rate: Money | None
    status: str | None
    data_quality_status: Quality
    warning_count: int = 0
    divergence_count: int = 0
    funding_status: Literal[
        "NOT_INFORMED",
        "INCOMPLETE",
        "COMPLETE",
        "OVERFUNDED",
        "BASE_AMOUNT_UNAVAILABLE",
    ] = "NOT_INFORMED"
    funding_identified_amount: Money = Decimal("0.00")
    funding_difference: Money | None = None
    funding_source_count: int = 0
    bank_validation_status: str = "NOT_RECORDED"
    continuity_id: str | None = None
    continuity_type: Literal["REFINANCING", "RENEGOTIATION", "ROLLOVER"] | None = None
    continuity_role: Literal["PREDECESSOR", "SUCCESSOR"] | None = None
    predecessor_sale_id: str | None = None
    predecessor_contract_code: str | None = None
    predecessor_sale_ids: list[str] = Field(default_factory=list)
    predecessor_contract_codes: list[str] = Field(default_factory=list)
    successor_sale_id: str | None = None
    successor_contract_code: str | None = None
    continuity_effective_date: date | None = None
    continuity_notes: str | None = None


class SaleDetail(SaleItem):
    warnings: list[QualityMessage] = Field(default_factory=list)
    divergences: list[QualityMessage] = Field(default_factory=list)


class SalesPage(OperationalApiModel):
    items: list[SaleItem]
    pagination: PageMeta
    summary: SaleSummary


class RevenueSummary(OperationalApiModel):
    total_records: int
    expected_amount: Money
    paid_amount: Money
    principal_received: Money
    interest_amount: Money
    discount_amount: Money
    pending_records: int
    warning_records: int
    divergent_records: int
    principal_total: Money = Decimal("0.00")
    principal_open: Money = Decimal("0.00")
    average_pmt: Money = Decimal("0.00")
    overdue_principal: Money = Decimal("0.00")
    overdue_pmt: Money = Decimal("0.00")
    delinquency_percentage: Money = Decimal("0.00")


class RevenueItem(OperationalApiModel):
    id: int
    revenue_identity_id: UUID | None = None
    contract_code: str | None
    client_name: str | None
    client_name_source: Literal["CLIENT_CANONICAL", "ECON_EMPRESTIMOS"] | None = None
    client_name_divergent: bool = False
    installment_code: str | None
    due_date: date | None
    payment_date: date | None
    expected_amount: Money | None
    paid_amount: Money | None
    principal_component: Money | None
    interest_component: Money | None
    discount_amount: Money | None
    installment_status: str | None
    situation: str | None
    anticipation_marker: str | None
    data_quality_status: Quality
    warning_count: int = 0
    divergence_count: int = 0
    sale_id: str | None = None
    funding_status: (
        Literal[
            "NOT_INFORMED",
            "INCOMPLETE",
            "COMPLETE",
            "OVERFUNDED",
            "BASE_AMOUNT_UNAVAILABLE",
        ]
        | None
    ) = None
    distribution_status: Literal[
        "PENDING_FUNDING",
        "READY",
        "DISTRIBUTED",
        "DIVERGENT",
        "REVERSED",
    ] = "PENDING_FUNDING"
    primary_source_name: str | None = None
    bank_validation_status: str = "NOT_RECORDED"
    refinanced_to_contract_code: str | None = None


class RevenueDetail(RevenueItem):
    payment_marker: str | None
    source_reference: str | None
    warnings: list[QualityMessage] = Field(default_factory=list)
    divergences: list[QualityMessage] = Field(default_factory=list)
    bank_validation_status: str = "NOT_RECORDED"


class RevenuePage(OperationalApiModel):
    items: list[RevenueItem]
    pagination: PageMeta
    summary: RevenueSummary
